### PR TITLE
Added .history.get() to the 13 endpoints that can return history, and…

### PR DESCRIPTION
**Pull request by brucem1976**
… updated integrations tests accordingly.

### CODE REVIEW

File: src/AccountingAPIClient.ts [hunk 1]
```diff
@@ -126,7 +126,20 @@ export class AccountingAPIClient extends BaseAPIClient {
 			const endpoint = 'banktransactions' + generateQueryString(args, true);
 			return this.oauth1Client.post<BankTransactionsResponse>(endpoint, bankTransaction);
 		},
-		attachments: this.generateAttachmentsEndpoint('banktransactions')
+		attachments: this.generateAttachmentsEndpoint('banktransactions'),
+		history: {
+			get: async (args?: { BankTransactionID: string }): Promise<HistoryResponse> => {
```
Comment (philals) [inline src/AccountingAPIClient.ts]: Just looking at the docs for H&N. I think the ID Guid is required. 

What do you think about making `args` required in all these endpoints? Aka remove the `?`

Comment (philals) [reply src/AccountingAPIClient.ts]: Eg: 

`get: async (args?: { BankTransactionID: string }): Promise<HistoryResponse> => {`
becomes:
`get: async (args: { BankTransactionID: string }): Promise<HistoryResponse> => {`

**Comment by brucem1976** [comment]
See my comments on the main thread - other tests are failing - but I ran the 13 individual tests for the endpoints that support .history.get() and they all worked

File: src/AccountingAPIClient.ts (hunk 1 above)
Comment (brucem1976) [reply src/AccountingAPIClient.ts]: Yes - that's correct - I've not used TypeScript before so I missed that. Shall I change and resubmit?

Comment (philals) [reply src/AccountingAPIClient.ts]: No worries.

Yup, just make the changes and push the new commit to your `master` the pull request will update automatically.

Comment (brucem1976) [reply src/AccountingAPIClient.ts]: Done

**Comment by brucem1976** [comment]
> Looking good. Just one comment below. Just thinking about a unit test, something like these ones: https://github.com/XeroAPI/xero-node/blob/master/src/__tests__/AccountingAPI-endpoint.tests.ts But will think about that a bit more.

Agree - I think this could be good.

**Comment by philals** [comment]
Sorry about the delay. I have been thinking about the best way to get around the rate limit issue.

I'm thinking what we do is move all these tests to the unit level. Something like https://github.com/XeroAPI/xero-node/blob/master/src/__tests__/AccountingAPI-attachments.tests.ts

We know that it will pass in CI as long as we don't add any more call, so it expedite your code I think we should do this.

Thoughts?

**Comment by brucem1976** [comment]
Agree!

**Comment by philals** [comment]
Here is what I have so far: 

```
import { AccountingAPIClient } from '../AccountingAPIClient';
import { XeroClientConfiguration } from '../internals/BaseAPIClient';
import { OAuth1HttpClient } from '../internals/OAuth1HttpClient';
import { InMemoryOAuthLibFactoryFactory } from '../internals/__tests__/helpers/InMemoryOAuthLib';
import { validTestCertPath } from '../internals/__tests__/helpers/privateKey-helpers';
import { mapConfig, mapState } from '../internals/config-helper';

interface IHistoryFixture {
	[key: string]: { idKey: string };
}

describe('AccountingAPI endpoints', () => {
	const guid1 = 'dcb417fc-0c23-4ba3-bc7f-fbc718e7e663';

	const xeroConfig: XeroClientConfiguration = {
		appType: 'private',
		consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
		consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
		privateKeyPath: validTestCertPath()
	};

	const inMemoryOAuthLibFF = new InMemoryOAuthLibFactoryFactory();

	const oauthConfig = mapConfig(xeroConfig, { apiBasePath: '/api.xro/2.0/' });
	const accountingBaseUrl = oauthConfig.apiBaseUrl;
	const oauthHttpClient = new OAuth1HttpClient(oauthConfig, mapState(xeroConfig), inMemoryOAuthLibFF.newFactory());
	const xeroClient = new AccountingAPIClient(xeroConfig, null, oauthHttpClient);

	const fixtures: IHistoryFixture = {
		prepayments: { idKey: 'PrepaymentID' }
	};

	Object.keys(fixtures).map((endpoint: string) => {
		Object.keys(fixtures[endpoint]).map((fixture) => {

			describe(`${endpoint} and getting History`, () => {
				let result: any;

				const mockedResponse = JSON.stringify({ a: 'response' });

				beforeAll(async () => {
					inMemoryOAuthLibFF.inMemoryOAuthLib.reset();
					inMemoryOAuthLibFF.inMemoryOAuthLib.setResponse(false, mockedResponse, { statusCode: 200 });

					const args: any = {};
					args[fixture] = guid1;

					result = await (xeroClient as any)[endpoint][fixture]['history'](args);
				});

				it(`calls the ${endpoint} url`, () => {
					inMemoryOAuthLibFF.inMemoryOAuthLib.lastCalledThisURL(accountingBaseUrl + '/api.xro/2.0/' + endpoint + '/history');
				});

				it('matches the expected response', () => {
					expect(result).toMatchObject(JSON.parse(mockedResponse));
				});
			});
		});

	});
});

```

Will continue. 

It was actually easier to copy from https://github.com/XeroAPI/xero-node/blob/master/src/__tests__/AccountingAPI-endpoint.tests.ts

**Comment by philals** [comment]
Sorry - I gotta head off. Will maybe be able to get some time in the weekend.

Here is what I have at the moment, feel free to see if you can finish it:

```import { AccountingAPIClient } from '../AccountingAPIClient';
import { XeroClientConfiguration } from '../internals/BaseAPIClient';
import { OAuth1HttpClient } from '../internals/OAuth1HttpClient';
import { InMemoryOAuthLibFactoryFactory } from '../internals/__tests__/helpers/InMemoryOAuthLib';
import { validTestCertPath } from '../internals/__tests__/helpers/privateKey-helpers';
import { mapConfig, mapState } from '../internals/config-helper';

interface IHistoryFixture {
	[key: string]: { idKey: string };
}

describe('AccountingAPI endpoints', () => {
	const guid1 = 'dcb417fc-0c23-4ba3-bc7f-fbc718e7e663';

	const xeroConfig: XeroClientConfiguration = {
		appType: 'private',
		consumerKey: 'RDGDV41TRLQZDFSDX96TKQ2KRJIW4C',
		consumerSecret: 'DJ3CMGDB0DIIA9DNEEJMRLZG0BWE7Y',
		privateKeyPath: validTestCertPath()
	};

	const inMemoryOAuthLibFF = new InMemoryOAuthLibFactoryFactory();

	const oauthConfig = mapConfig(xeroConfig, { apiBasePath: '/api.xro/2.0/' });
	const accountingBaseUrl = oauthConfig.apiBaseUrl;
	const oauthHttpClient = new OAuth1HttpClient(oauthConfig, mapState(xeroConfig), inMemoryOAuthLibFF.newFactory());
	const xeroClient = new AccountingAPIClient(xeroConfig, null, oauthHttpClient);

	const fixtures: IHistoryFixture = {
		prepayments: { idKey: 'PrepaymentID' }
	};

	Object.keys(fixtures).map((endpoint: string) => {

		const idValue = fixtures[endpoint].idKey;

		describe(`${endpoint} and getting History`, () => {
			let result: any;

			const mockedResponse = JSON.stringify({ a: 'response' });

			beforeAll(async () => {
				inMemoryOAuthLibFF.inMemoryOAuthLib.reset();
				inMemoryOAuthLibFF.inMemoryOAuthLib.setResponse(false, mockedResponse, { statusCode: 200 });

				const args: any = {};
				args[idValue] = guid1;

				result = await (xeroClient as any)[endpoint]['History'](args);
			});

			it(`calls the ${endpoint} url`, () => {
				inMemoryOAuthLibFF.inMemoryOAuthLib.lastCalledThisURL(accountingBaseUrl + '/api.xro/2.0/' + endpoint + '/history');
			});

			it('matches the expected response', () => {
				expect(result).toMatchObject(JSON.parse(mockedResponse));
			});
		});
	});

});
```

**Comment by brucem1976** [comment]
Thanks, I am away until Sunday, will see if I can squeeze it in somewhere

**Pull request merged**